### PR TITLE
fix: toxsave memory leak while loading groups

### PR DESCRIPTION
**Pull request by sudden6**
Found here: https://github.com/TokTok/c-toxcore/actions/runs/1863916365

Depends on: https://github.com/TokTok/c-toxcore/pull/2058

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2064)

### CODE REVIEW

File: toxcore/group.c [hunk 1]
```diff
@@ -3392,116 +3392,166 @@ uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data)
...
+
 non_null()
-static State_Load_Status load_conferences(Group_Chats *g_c, const uint8_t *data, uint32_t length)
+static uint32_t load_group(Group_c *g, const Group_Chats *g_c, const uint8_t *data, uint32_t length)
 {
     const uint8_t *init_data = data;
 
-    while (length >= (uint32_t)(data - init_data) + SAVED_CONF_SIZE_CONSTANT) {
-        const int groupnumber = create_group_chat(g_c);
+    // Initialilize to default values so we can unconditionally free in case of an error
```
Comment (iphydf) [inline toxcore/group.c:3409]: "Initialize"

File: toxcore/group.c [hunk 2]
```diff
@@ -3392,116 +3392,166 @@ uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data)
...
 
-            id_copy(peer->real_pk, data);
-            data += CRYPTO_PUBLIC_KEY_SIZE;
-            id_copy(peer->temp_pk, data);
-            data += CRYPTO_PUBLIC_KEY_SIZE;
+    id_copy(g->real_pk, nc_get_self_public_key(g_c->m->net_crypto));
 
-            lendian_bytes_to_host16(&peer->peer_number, data);
-            data += sizeof(uint16_t);
+    return (uint32_t)(data - init_data);
```
Comment (iphydf) [inline toxcore/group.c:3500]: Are all these casts safe? Do we need asserts to make sure they are?

File: toxcore/group.c [hunk 3]
```diff
@@ -3392,116 +3392,166 @@ uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data)
...
+        data += consumed;
 
-        g->status = GROUPCHAT_STATUS_CONNECTED;
-        memcpy(g->real_pk, nc_get_self_public_key(g_c->m->net_crypto), CRYPTO_PUBLIC_KEY_SIZE);
         const int peer_index = addpeer(g_c, groupnumber, g->real_pk, dht_get_self_public_key(g_c->m->dht), g->peer_number,
                                        nullptr, true, false);
 
         if (peer_index == -1) {
-            return STATE_LOAD_STATUS_ERROR;
+            goto LOAD_ERROR;
```
Comment (iphydf) [inline toxcore/group.c:3536]: Let's try to avoid adding more `goto` and instead use an inner function with a cleanup outside.

**Review by iphydf — CHANGES_REQUESTED** [review]

File: toxcore/group.c (hunk 2 above)
Comment (sudden6) [reply toxcore/group.c:3508]: I just took this from the existing code, but added asserts anyway.

File: toxcore/group.c (hunk 3 above)
Comment (sudden6) [reply toxcore/group.c]: Changed the structure, though no inner function.

Comment (sudden6) [reply toxcore/group.c]: Finally understand how you meant that and did it.

File: toxcore/group.c [hunk 4]
```diff
@@ -3504,6 +3552,27 @@ static State_Load_Status load_conferences(Group_Chats *g_c, const uint8_t *data,
...
+    State_Load_Status res = load_conferences_helper(g_c, data, length);
+
+    if (res == STATE_LOAD_STATUS_CONTINUE) {
+        return res;
+    }
+
+    // Loading failed, cleanup all Group_c
+
+    // save locally, because wipe_group_chat(...) modifies it
+    uint16_t num_groups = g_c->num_chats;
```
Comment (iphydf) [inline toxcore/group.c:3567]: const

File: toxcore/group.c [hunk 5]
```diff
@@ -3504,6 +3552,27 @@ static State_Load_Status load_conferences(Group_Chats *g_c, const uint8_t *data,
     return STATE_LOAD_STATUS_CONTINUE;
 }
 
+non_null()
+static State_Load_Status load_conferences(Group_Chats *g_c, const uint8_t *data, uint32_t length)
+{
+    State_Load_Status res = load_conferences_helper(g_c, data, length);
```
Comment (iphydf) [inline toxcore/group.c:3558]: const

File: toxcore/group.c [hunk 6]
```diff
@@ -3504,6 +3552,27 @@ static State_Load_Status load_conferences(Group_Chats *g_c, const uint8_t *data,
...
+    if (res == STATE_LOAD_STATUS_CONTINUE) {
+        return res;
+    }
+
+    // Loading failed, cleanup all Group_c
+
+    // save locally, because wipe_group_chat(...) modifies it
+    uint16_t num_groups = g_c->num_chats;
+
+    for (uint16_t i = 0; i < num_groups; i++) {
```
Comment (iphydf) [inline toxcore/group.c:3569]: Fix cimple errors (`++i`).

File: toxcore/group.c [hunk 7]
```diff
@@ -3392,105 +3392,153 @@ uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data)
...
+        if (groupnumber == -1) {
+            // If this fails there's a serious problem, don't bother with cleanup
+            return STATE_LOAD_STATUS_ERROR;
         }
 
-        if (g->numfrozen > g->maxfrozen) {
-            g->maxfrozen = g->numfrozen;
+        Group_c *g = &g_c->chats[groupnumber];
+
+        uint32_t consumed = load_group(g, g_c, data, length - (uint32_t)(data - init_data));
```
Comment (iphydf) [inline toxcore/group.c:3529]: const

File: toxcore/group.c [hunk 8]
```diff
@@ -3392,105 +3392,153 @@ uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data)
...
 
+/**
+ * @brief load_group Load a Group section from a save file
+ * @param g Group to load
+ * @param g_c Reference to all groupchats, need for utility functions
+ * @param data Start of the data to deserialze
+ * @param length Length of data
+ * @return 0 on error, number of consumed bytes otherwise
+ */
+
```
Comment (iphydf) [inline toxcore/group.c:3403]: nit: remove empty line between comment and thing-to-be-commented.

**Review by iphydf — APPROVED** [review]

**Pull request merged**